pkg/pillar/nistate/linux_flow.go

```diff
@@ -417,22 +463,67 @@ func (lc *LinuxCollector) processCapturedPacket(
 	}
 	packet := capPacket.packet
 	dnslayer := packet.Layer(layers.LayerTypeDNS)
-	if niInfo.config.Type == types.NetworkInstanceTypeSwitch && dnslayer == nil {
-		addrUpdates, isDhcp := lc.processDHCPPacket(niInfo, packet)
-		if isDhcp {
+	if packet.NetworkLayer() != nil {
+		if niInfo.config.Type == types.NetworkInstanceTypeSwitch && dnslayer == nil {
+			addrUpdates, isDhcp := lc.processDHCPPacket(niInfo, packet)
+			if isDhcp {
+				return addrUpdates
+			}
+			return lc.processDADProbe(niInfo, packet)
+		}
+		if dnslayer != nil {
+			lc.processDNSPacketInfo(niInfo, packet)
+		}
+	} else if niInfo.config.Type == types.NetworkInstanceTypeSwitch && packet.LinkLayer() != nil {
+		addrUpdates, isArp := lc.processARPPacket(niInfo, packet)
+		if isArp {
 			return addrUpdates
 		}
-		return lc.processDADProbe(niInfo, packet)
-	}
-	if dnslayer != nil {
-		lc.processDNSPacketInfo(niInfo, packet)
 	}
 	return nil
 }
 
 // Used as a constant.
 var broadcastMAC = []byte{0xff, 0xff, 0xff, 0xff, 0xff, 0xff}
 
+func (lc *LinuxCollector) processARPPacket(
+	niInfo *niInfo, packet gopacket.Packet) (addrUpdates []VIFAddrsUpdate, isARP bool) {
+
+	arpLayer := packet.Layer(layers.LayerTypeARP)
+	if arpLayer == nil {
+		return nil, false
+	}
+
+	arp, _ := arpLayer.(*layers.ARP)
+	if arp == nil {
+		return nil, false
+	}
+
```

Comment:
Below this point you can return `isARP` as `true` - however, currently it is irrelevant since there is no other packet processing after this function

Reply:
yes, but the 'real' meaning is we got an 'arp IP address'. We'll see later when there is more items

Reply:
In that case you can drop `isARP` and just do `len(addrUpdates) != 0` instead.
In `processDHCPPacket` the boolean is used to decide if after `processDHCPPacket` we should try to run `processDADProbe`. Because even if `processDHCPPacket` learned nothing about IP assignments but confirmed that the packet is DHCP, there is still no point trying to run `processDADProbe`.

Reply:
sure.